File: ext/examples/resolve_through_stem.rs
//! Resolves a module up to an $(n, s)$ and prints an ASCII depiction of the Ext groups:
//! ```text
//! ·
//! ·                                                     ·
//! ·                                                   · ·     ·
//! ·                                                 ·   ·     ·
//! ·                                             ·   ·         ·
//! ·                                     ·       · · ·         ·
//! ·                                   · ·     · · · ·     ·   ·
//! ·                                 ·   ·     · :   · ·   · · ·
//! ·                             ·   ·         · ·   · ·   · · ·
//! ·                     ·       · · ·           ·     ·     · ·
//! ·                   · ·     · · · ·     ·     ·     ·       ·
//! ·                 ·   ·     · :   · ·   · ·   · ·           ·
//! ·             ·   ·         · ·   · :   ·   · ·             ·
//! ·     ·       · · ·         · ·   · · ·   ·                 ·
//! ·   · ·     · · ·           · · ·   ·                       ·
//! · ·   ·       ·               ·
//! ·
//! ```

use ext::chain_complex::FreeChainComplex;
use sseq::coordinates::Bidegree;

fn main() -> anyhow::Result<()> {
    let res = ext::utils::query_module_only("Module", None, false)?;

    let max = Bidegree::n_s(
        query::with_default("Max n", "30", str::parse),
        query::with_default("Max s", "15", str::parse),
    );

    res.compute_through_stem(max);

    println!("{}", res.graded_dimension_string());

    Ok(())
}

File: ext_py/examples/resolve_through_stem.py
#!/usr/bin/env python3
"""Resolve a module up to an (n, s) and print an ASCII depiction of Ext.

Python port of ext/examples/resolve_through_stem.rs.
"""

import _query as query
from ext import sseq


def main():
    res = query.query_resolution("Module")

    max_bidegree = sseq.Bidegree.n_s(
        query.with_default("Max n", "30", int),
        query.with_default("Max s", "15", int),
    )

    res.compute_through_stem(max_bidegree)

    print(res.graded_dimension_string())


if __name__ == "__main__":
    main()
